pkg/monitoring/endpoint.go

```diff
@@ -0,0 +1,389 @@
+	_, err := w.Write(*data)
+	if err != nil {
+		http.Error(w, "failed to write response: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func setLongerPrefixes(r *http.Request, command *[]string) error {
+	longerPrefixes := r.URL.Query().Get("longer_prefixes")
+	if longerPrefixes != "" {
+		useLongerPrefixes, err := strconv.ParseBool(longerPrefixes)
+		if err != nil {
+			return fmt.Errorf("longer_prefixes value '%s' is not valid: %w", longerPrefixes, err)
+		}
+		if useLongerPrefixes {
+			*command = append(*command, "longer-prefixes")
+		}
+	}
+	return nil
+}
+
+func setInput(r *http.Request, command *[]string) error {
+	input := r.URL.Query().Get("input")
+	if input != "" {
+		if _, _, err := net.ParseCIDR(input); err != nil {
+			return fmt.Errorf("input '%s' is not valid: %w", input, err)
+		}
+		*command = append(*command, input)
+	}
+	return nil
+}
+
+func withNodename(data *[]byte) (*[]byte, error) {
+	res := map[string]json.RawMessage{}
+	nodename := os.Getenv(healthcheck.NodenameEnv)
+
+	result := data
+	if nodename != "" {
+		res[nodename] = json.RawMessage(*data)
+		var err error
+		*result, err = json.MarshalIndent(res, "", "\t")
+		if err != nil {
+			return nil, fmt.Errorf("error marshaling data: %w", err)
+		}
+	}
+
+	return result, nil
+}
+
+func passRequest(r *http.Request, addr, query string, results chan []byte, errors chan error, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	s := strings.Split(r.Host, ":")
+	port := ""
+	if len(s) > 1 {
+		port = s[1]
+	}
+
+	protocol := "http"
+	if r.TLS != nil {
+		protocol = "https"
+	}
+
+	url := fmt.Sprintf("%s://%s:%s%s", protocol, addr, port, query)
+	resp, err := http.Get(url) //nolint
+	if err != nil {
+		errors <- fmt.Errorf("error getting data from %s: %w", addr, err)
+		return
+	}
+
+	data, err := io.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		errors <- fmt.Errorf("error reading response from %s: %w", addr, err)
+		return
+	}
+
+	results <- data
+}
+
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=list
+
+func (e *Endpoint) getAddresses(ctx context.Context, svc *corev1.Service) ([]string, error) {
+	var serviceLabels labels.Set = svc.Spec.Selector
+	pods := &corev1.PodList{}
+	if err := e.c.List(ctx, pods, &client.ListOptions{
+		LabelSelector: serviceLabels.AsSelector(),
+		Namespace:     svc.Namespace,
+	}); err != nil {
+		return nil, fmt.Errorf("error getting pods: %w", err)
+	}
+
+	addresses := []string{}
+	for i := range pods.Items {
+		addresses = append(addresses, pods.Items[i].Status.PodIP)
+	}
+
+	return addresses, nil
+}
+
+func queryEndpoints(r *http.Request, addr []string) ([]byte, error) {
+	query := strings.ReplaceAll(r.URL.RequestURI(), "all/", "")
+	responses := []json.RawMessage{}
+
+	var wg sync.WaitGroup
+	results := make(chan []byte, len(addr))
+	errors := make(chan error, len(addr))
+
+	for i := range addr {
+		wg.Add(1)
+		go passRequest(r, addr[i], query, results, errors, &wg)
```

Comment:
You could also do this so it's a little bit more separated
```suggestion
		go func() {
		  passRequest(r, addr[i], query, results, errors, &wg)
		  wg.Done()
		}()
```

Reply:
Added that. However had to change it slightly as it seems that `i` variable is a reference, so decided to pass it by value to the lambda function just to be sure it won't get changed outside the function.